frontend/src/components/editor/package-alert.tsx

```diff
@@ -0,0 +1,252 @@
+                    href="https://docs.python.org/3/tutorial/venv.html#creating-virtual-environments"
+                    className="text-accent-foreground hover:underline"
+                    target="_blank"
+                    rel="noreferrer"
+                  >
+                    virtual environment
+                  </a>
+                  , marimo can install these packages for you.
+                </p>
+              )}
+            </div>
+          </div>
+        </Banner>
+      </div>
+    );
+  } else if (isInstallingPackageAlert(packageAlert)) {
+    const { status, title, titleIcon, description } =
+      getInstallationStatusElements(packageAlert.packages);
+    if (status === "installed") {
+      setTimeout(() => clearPackageAlert(packageAlert.id), 10_000);
+    }
+
+    return (
+      <div className="flex flex-col gap-4 mb-5 fixed top-5 left-5 w-[400px] z-[200] opacity-95">
+        <Banner
+          kind={status === "failed" ? "danger" : "info"}
+          className="flex flex-col rounded pt-3 pb-4 px-5"
+        >
+          <div className="flex justify-between">
+            <span className="font-bold text-lg flex items-center mb-2">
+              {titleIcon}
+              {title}
+            </span>
+            <Button
+              variant="text"
+              size="icon"
+              onClick={() => clearPackageAlert(packageAlert.id)}
+            >
+              <XIcon className="w-5 h-5" />
+            </Button>
+          </div>
+          <div
+            className={cn(
+              "flex flex-col gap-4 justify-between items-start text-muted-foreground text-base",
+              status === "installed" && "text-accent-foreground",
+            )}
+          >
+            <div>
+              <p>{description}</p>
+              <ul className="list-disc ml-4 mt-1">
+                {Object.entries(packageAlert.packages).map(
+                  ([pkg, st], index) => (
+                    <li
+                      className={cn(
+                        "flex items-center gap-1 font-mono text-sm",
+                        st === "installing" && "font-semibold",
+                        st === "failed" && "text-destructive",
+                        st === "installed" && "text-accent-foreground",
+                        st === "installed" &&
+                          status === "failed" &&
+                          "text-muted-foreground",
+                      )}
+                      key={index}
+                    >
+                      <ProgressIcon status={st} />
+                      {pkg}
+                    </li>
+                  ),
+                )}
+              </ul>
+            </div>
+          </div>
+        </Banner>
+      </div>
+    );
+  } else {
+    logNever(packageAlert);
+    return null;
+  }
+};
+
+function getInstallationStatusElements(packages: PackageInstallationStatus) {
+  const statuses = Object.entries(packages).map(([_, status]) => status);
```

Comment:
perf doesn matter here, but since you are just checking existence, you could use `new Set(Object.entries(packages).map(([_, status]) => status))`